pretraining/utils/dataset_utils.py

```diff
@@ -908,7 +928,10 @@ def __init__(
         # Build subdataset iterators
         self.data = []
         for i, d in enumerate(self.datasets):
-            self.data.append(dataset(**passthrough_args, datasets=[d]))
+            stateful_dataset = dataset(**passthrough_args)
```

Comment:
the kwarg is removed from the dataset function here?

Reply:
Only `datasets` because it does exist in the `__init__` function of a `_Stateful_Dataset`, but it's added in L933.
I know it may not be identical as the `__init__` function may do something different with `datasets`, the alternative is to change the `__init__` of `_Stateful_Dataset` to include `datasets` there

Reply:
I think the annotation on `dataset` is wrong so you're assuming a type is passed in that doesn't accept a datasets parameter, but actually a different type is passed in that does. it looks like this is the actual call-site:

https://github.com/foundation-model-stack/pretrain/blob/2e76f9919bed17583745f6cbb634394a0f96bad1/pretraining/utils/dataloader_utils.py#L54

Reply:
I found a better solution, update coming

Reply:
This was `mypy` original error:

```
pretraining/utils/dataset_utils.py:911: error: Unexpected keyword argument "datasets" for "_Stateful_Dataset"  [call-arg]
```

That's because from the typing information available to `mypy` there was no guarantee that the `dataset` in `self.data.append(dataset(**passthrough_args, datasets=[d]))` would have a `datasets` keyword argument in its `__init__`. 

The new protocol `_Stateful_Dataset_With_Datasets` added lets us tell `mypy` that we expect a class with a `datasets` argument in its constructor. It also prevents in future someone from inadvertently passing a `dataset` object that does not fit.